ostd/src/arch/riscv/trap/mod.rs

```diff
@@ -29,51 +31,90 @@ pub(crate) unsafe fn init() {
-                        InterruptSource::Timer,
-                    ),
-                    PrivilegeLevel::Kernel,
-                );
-            }
-            Interrupt::SupervisorExternal => {
-                // No races because we are in IRQs.
-                let current_cpu = CpuId::current_racy().into();
-                while let Some(hw_irq_line) = IRQ_CHIP.get().unwrap().claim_interrupt(current_cpu) {
-                    call_irq_callback_functions(f, &hw_irq_line, PrivilegeLevel::Kernel);
+    fn disable_local_if(cond: bool) {
+        if cond {
+            disable_local();
+        }
+    }
+
+    use riscv::register::scause::Trap::*;
+
+    let scause = riscv::register::scause::read();
+    match scause.cause() {
+        Interrupt(interrupt) => {
+            use riscv::register::scause::Interrupt::*;
+
+            match interrupt {
+                SupervisorTimer => {
+                    call_irq_callback_functions(
+                        f,
+                        &HwIrqLine::new(
+                            TIMER_IRQ_NUM.load(Ordering::Relaxed),
+                            InterruptSource::Timer,
+                        ),
+                        PrivilegeLevel::Kernel,
+                    );
                 }
-            }
-            Interrupt::SupervisorSoft => todo!(),
-            _ => {
-                panic!(
-                        "cannot handle unknown supervisor interrupt: {interrupt:?}. trapframe: {f:#x?}.",
+                SupervisorExternal => {
+                    // No races because we are in IRQs.
+                    let current_cpu = CpuId::current_racy().into();
+                    while let Some(hw_irq_line) =
+                        IRQ_CHIP.get().unwrap().claim_interrupt(current_cpu)
+                    {
+                        call_irq_callback_functions(f, &hw_irq_line, PrivilegeLevel::Kernel);
+                    }
+                }
+                SupervisorSoft => todo!(),
+                Unknown => {
+                    panic!(
+                        "Cannot handle unknown supervisor interrupt, scause: {:#x}, trapframe: {:#x?}.",
+                        scause.bits(), f
                     );
+                }
             }
-        },
-        Trap::Exception(e) => {
-            let stval = riscv::register::stval::read();
-            panic!(
-                "Cannot handle kernel cpu exception: {e:?}. stval: {stval:#x}, trapframe: {f:#x?}.",
-            );
+        }
+        Exception(e) => {
+            use CpuException::*;
+
+            let exception = e.into();
+            // The IRQ state before trapping. We need to ensure that the IRQ state
+            // during exception handling is consistent with the state before the trap.
+            let was_irq_enabled = riscv::register::sstatus::read().spie();
+            enable_local_if(was_irq_enabled);
+            match exception {
+                Unknown => {
```

Comment:
Following my previous comments, I think we can do this:
```rust
            let scause = riscv::register::scause::read();
            let exception = match scause.cause() {
                Trap::Interrupt(interrupt) => {
                    call_irq_callback_functions_by_scause(interrupt);
                    return;
                },
                Trap::Exception(raw_exception) => {
                    exception = CpuException::new(raw_exception, stval);
                    enable_local_if(was_irq_enabled);
                },
             };

             match exception {
                 // ...
             }

            disable_local_if(was_irq_enabled);
```

So the code (without the nested `match`) looks better?